### PR TITLE
Set up associations

### PR DESCRIPTION
**Pull request by rbanos-mv**
# OOP school library

> Imagine that you are the librarian of OOP University, and you need a tool to record what books are in the library and who borrows them. The app that you will create will allow you to:

- Add new students or teachers.
- Add new books.
- Save records of who borrowed a given book and when.

And all of this will be built in a beautiful and well-organized way!

### See the decorators in action
```
irb oop_school_library.rb
```

### Project requirements
- Create a class `Classroom` with the following:
  - `@label` instance variable, should be initialized in the constructor.
  - Setter and getter for `@label` (remember about `attr_accessor`).
- Create the `has-many`/`belongs-to` relationship between `Classroom` and `Student`. The following should be implemented:
  - Create the `has-many` side (a classroom has many students).
  - Create the `belongs-to` side (a student belongs to a classroom).
  - Make sure that when adding a student to a classroom it also sets the classroom for the student.
  - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- Create a class `Book` with the following:
  - `@title` and `@author` instance variables, should be initialized in the constructor.
  - Setters and getters for instance variables (remember about `attr_accessor`).
- Create a class `Rental` with the following:
  - `@date` instance variable, should be initialized in the constructor.
  - Setter and getter for `@date` (remember about `attr_accessor`).
- Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following should be implemented:
  - Create the `has-many` side of `Book` and `Rental` (a book has many rentals).
  - Create the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
  - Create the `has-many` side of `Person` and `Rental` (a person has many rentals).
  - Create the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
  - Modify the constructor of `Rental` so `Book` and `Person` are set in it.

#### Interface
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.

#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### General requirements

- Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

### Ruby requirements
- Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### Project requirements
- Create class `Person` with  the following:
  - Instance vars: `@id`, `@name`, and `@age`.
  - Constructor with `name`,  `age`, and `parent_permission` as parameter. `name` and `parent_permission` are optional and have default values of `"Unknown"` and `true`.
  - Getters for `@id`, `@name`, and `@age`.
  - Setters for `@name` and `@age`.
  - Private method `is_of_age?` that returns `true` if `@age` is greater or equal to 18 and `false` otherwise.
  - Public method `can_use_services?` that returns `true` if person is of age or if they have permission from parents.
- Create class `Student` with the following:
  - Inherits from `Person`.
  - Constructor extends parent's constructor by adding `@classroom` and a parameter for it.
  - Method `play_hooky` that returns `"¯\(ツ)/¯"`.
- Create class `Teacher` with the following:
  - Inherits from `Person`.
  - Constructor extends parent's constructor by adding `@specialization` and a parameter for it.
  - Override `can_use_services?` so it always returns `true`.
- Each class should be saved in a separate file.

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Classroom
+  attr_accessor :label
+  attr_reader :students
+
+  def initialize
+    @label = ''
+    @students = []
+  end
```
Comment (BenMukebo) [inline classroom.rb]: - [ ] Great bob 👏🏻  Let's modify our constructor to accept parameters and use them. I might be like this:

```
class Animal
  def initialize(name, number_of_legs)
    @id = Random.rand(1..1000)
    @name = name
    @number_of_legs = number_of_legs
  end
end
```

Visit this [link](`https://github.com/rbanos-mv/oop_school_library/pull/3/files#diff-bd6d401bca7b47b397906cb44f68b377a8e8570fed0f0d54848131cf14baa70d`)

File: classroom.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  def add_student(student)
+    @students.push(student) unless @students.include?(student)
```
Comment (BenMukebo) [inline classroom.rb]: - [ OPTIONAL] You can use `<<` symbole instead .push key word. eg: `@students << student`

File: person.rb [hunk 3]
```diff
@@ -20,6 +21,10 @@ def can_use_services?
     of_age? || parent_permission
   end
 
+  def add_rental(rental)
+    @rentals.push(rental) unless @rentals.include?(rental)
+  end
```
Comment (BenMukebo) [inline person.rb]: - [ ] Same here 🤔 what if we add the date?

```ruby
 def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

**Review by BenMukebo — CHANGES_REQUESTED** [review]
Hi @rbanos-mv, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

You have submitted a project, that Linters check are not running in the GitHub/action
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please Tag me on (@BenMukebo) when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by rbanos-mv** [comment]
Fixed required changes
- add parameters to initialize method in Classroom class
- refactor parameters to add_rentals in both Book and Person classes
- substitute .push with <<

**I don't understand why @BenMukebo claims that "linters are not running in the GitHub/action" when it can be verified in the history of actions that linters are in place and running.**
image
image

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @rbanos-mv👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by BenMukebo** [comment]
Hi @rbanos-mv, You don't have to worry about that. 😄  In order to save time, we had to save a few PR resume templates to overwrite each time you give the student a review. The thing is that 🤔  I forgot to remove the linker error message that was logged in the summary, but it was by mistake. If a code review leaves or gives a change required by mistake, please contact him or tag him here so he can see if he deliberately forgot this. Normally a PR that has linter errors is not allowed to review that. Instead, we mark it as invalid. Hope that is more clear now 👍🏻

**Pull request merged**